bazaarvoice/emodb
# block column functionality

## Github Issue #

None

## What Are We Doing Here?

EmoDB uses the CQL driver to query Cassandra. The CQL interface uses Cassandra native transport. This transport includes a maximum frame size which is 256MB by default. This means that any single CQL query cannot return more than 256MB in a single batch, although there is no limitation on the size of the query’s entire result set.

Right now, all CQL queries Emo makes results are returned in batches of 100 rows. There are several circumstances where this can exceed the maximum frame size, including:
* A single document has numerous large deltas which for one reason or another have not been compacted (e.g.; A document gets 100 3MB deltas in under 5 minutes)
* A table scan is run on a table with a large average document size (e.g.; Scan a table with 100 3MB compacted documents)

The solution for this is to split long delta values across multiple rows in C*. The primary key will include a third value, block, used for ordering the rows which contain a single change. For example, Emo can now enforce that each row’s value is at most 64 Kb. Additionally, a prefix will be added to the first block of every delta value that indicates how many blocks it consists of.

Writing a 150 kb delta would result in the following rows:

| rowkey        | changeid       | block  | value
| ------------|:-------------:| -----:| ------------:|
| coord1      | t1 | 0 | 0003<kilobytes (0...61)>
| coord1      | t1      |   1 | <kilobytes (61...125)>
| coord1 | t1      |    2 | <kilobytes (125...150)>

Using this method we can guarantee that emo can query with batch sizes of 4096 (256Mb / 64Kb) and never breach its maximum frame size.

The way I implemented this change is entirely at the DAO layer. Whenever a delta is passed to a WriterDAO to be written, it is first partitioned into blocks and then the blocks are together written to C* atomically. When a delta is read by a readerDAO, it is stitched together by the new DeltaIterator class. The logic of any layer of Emo above the DAO should remain unchanged, as they do not need to be aware of how deltas are stored in C*. 

One problem that blocking created was that it became impossible to delete compacted deltas using Thrift. This is because Thrift requires you to provide the entire primary key for every record you want to delete. To address this issue, I transitioned all compaction functionality to a new CqlDataWriterDAO. Using CQL, we can delete compacted deltas easily by doing `DELETE FROM <table> WHERE rowkey=<rowkey> AND changeid=<changeid>`. We are no longer required to provide the entire primary key. One major benefit to this change is that compacted deltas can now be larger than the 15MB maximum Thrift transport size and are now bound by the 256 MB CQL max transport size. 

## How to Test and Verify

1. Check out this PR
2. Build and run Emo locally
3. Write some deltas to Emo and then read them back out. To you, as a client, the functionality should be the exact same as it was prior to the change.
4. Run the cqlsh and peek at the tables. You should see this blocking in action.
5. Run gatekeeper and ensure that emo is still able to handle all requests the same as it could previously.

Note: In order to truly see the blocking functionality in action, you will probably have to significantly decrease the block size inside of the DAOModule. This is because you will probably find it difficult to write deltas more than 64kb.

## Risk

### Level 

`High`

### Required Testing

`Regression` and `Manual`

### Risk Summary

The risk here is pretty straightforward. I have modified the way in which data stored and read out of C*. If something isn't working correctly, it could be catastrophic.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
